backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/discussion/DiscussionService.java

```diff
@@ -0,0 +1,273 @@
+        if (discussion.getQuestion() == null ) {
+            throw new TutorException(DISCUSSION_MISSING_QUESTION);
+        }
+        if(discussion.getUserId() == null){
+            throw new TutorException(DISCUSSION_MISSING_USER);
+        }
+        if(discussion.getMessage() == null || discussion.getMessage().trim().length() == 0){
+            throw new TutorException(DISCUSSION_MISSING_MESSAGE);
+        }
+    }
+
+    private void checkUserAndQuestion(User user, Question question) {
+        if (user.getRole() == User.Role.TEACHER) {
+            throw new TutorException(DISCUSSION_NOT_STUDENT_CREATOR);
+        }
+
+        checkUserAnswered(user, question);
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<DiscussionDto> findDiscussionsByQuestionId(Integer questionId) {
+        return discussionRepository.findByQuestionId(questionId).stream().map(DiscussionDto::new)
+                .collect(Collectors.toList());
+    }
+
+    private void checkUserAnswered(User user, Question question) {
+        if (!user.checkQuestionAnswered(question)) {
+            throw new TutorException(QUESTION_NOT_ANSWERED, question.getId());
+        }
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<DiscussionDto> findDiscussionsByUserId(Integer userId) {
+        return discussionRepository.findByUserId(userId).stream().map(DiscussionDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<DiscussionDto> findDiscussionsByCourseExecutionId(int courseExecutionId)  {
+        return discussionRepository.findByCourseExecutionId(courseExecutionId).stream().map(DiscussionDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<DiscussionDto> findOpenDiscussionsByCourseExecutionId(int courseExecutionId)  {
+        return discussionRepository.findByCourseExecutionId(courseExecutionId).stream().filter(discussion -> !discussion.isClosed())
+                .map(DiscussionDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<DiscussionDto> getAnsweredDiscussions(int courseExecutionId, Integer userId) {
+        return discussionRepository.findByCourseExecutionIdAndUserId(courseExecutionId, userId).stream()
+                .filter(discussion -> (!discussion.getReplies().isEmpty()))
```

Comment:
idem, but courious why this is necessary

Reply:
It is used in the students stats page

Reply:
OK!